### ISSUE TITLE
query-frontend: No caching when used with query configured with prefix

### ISSUE THREAD

**Issue by gburek-fastly** [thanos-io/thanos#3178]
**Thanos, Prometheus and Golang version used**:

```
$ thanos --version
thanos, version 0.15.0 (branch: HEAD, revision: fbd14b49f95e7543883fedf9381a21127d4dda2b)
  build user:       root@d35b081d3e19
  build date:       20200907-09:29:21
  go version:       go1.14.2
```

**What happened**:

Query-frontend is not caching or otherwise modifying the queries that flow through it to downstream query nodes. This can be seen by incrementing `http_requests_*` and `http_responses_*` metrics but `querier_*` and `cortex_*` metrics remain at 0:

<details>Metrics
<p>

```
$ curl -s 127.0.0.1:9090/metrics | grep querier
# HELP querier_cache_added_new_total The total number of new entries added to the cache
# TYPE querier_cache_added_new_total counter
querier_cache_added_new_total{cache="fifocache"} 0
# HELP querier_cache_added_total The total number of Put calls on the cache
# TYPE querier_cache_added_total counter
querier_cache_added_total{cache="fifocache"} 0
# HELP querier_cache_entries The total number of entries
# TYPE querier_cache_entries gauge
querier_cache_entries{cache="fifocache"} 0
# HELP querier_cache_evicted_total The total number of evicted entries
# TYPE querier_cache_evicted_total counter
querier_cache_evicted_total{cache="fifocache"} 0
# HELP querier_cache_gets_total The total number of Get calls
# TYPE querier_cache_gets_total counter
querier_cache_gets_total{cache="fifocache"} 0
# HELP querier_cache_memory_bytes The current cache size in bytes
# TYPE querier_cache_memory_bytes gauge
querier_cache_memory_bytes{cache="fifocache"} 0
# HELP querier_cache_misses_total The total number of Get calls that had no valid entry
# TYPE querier_cache_misses_total counter
querier_cache_misses_total{cache="fifocache"} 0
# HELP querier_cache_stale_gets_total The total number of Get calls that had an entry which expired
# TYPE querier_cache_stale_gets_total counter
querier_cache_stale_gets_total{cache="fifocache"} 0

$ curl -s 127.0.0.1:9090/metrics | grep thanos_query_frontend_queries_total
# HELP thanos_query_frontend_queries_total Total queries passing through query frontend
# TYPE thanos_query_frontend_queries_total counter
thanos_query_frontend_queries_total{op="query"} 0
thanos_query_frontend_queries_total{op="query_range"} 0

$ curl -s 127.0.0.1:9090/metrics | grep http_requests_total
# HELP http_requests_total Tracks the number of HTTP requests.
# TYPE http_requests_total counter
http_requests_total{code="200",handler="query-frontend",method="get"} 12324
http_requests_total{code="301",handler="query-frontend",method="get"} 229
http_requests_total{code="400",handler="query-frontend",method="get"} 2
http_requests_total{code="404",handler="query-frontend",method="get"} 16
http_requests_total{code="499",handler="query-frontend",method="get"} 2
```

</p>
</details>

**What you expected to happen**: query-frontend caching and splitting queries that flow through it. 

**How to reproduce it (as minimally and precisely as possible)**:

1. Run thanos query (port 9092) and query-frontend (port 9093) together with cli args:

```
thanos query --http-address=127.0.0.1:9092 --web.route-prefix=thanos --web.external-prefix=thanos ...

thanos query-frontend --http-address=127.0.0.1:9093 --query-frontend.compress-responses --query-frontend.downstream-url=http://127.0.0.1:9092 --query-range.split-interval=24h --query-range.max-retries-per-request=5 --query-frontend.log_queries_longer_than=5s --query-range.response-cache-config-file=/etc/thanos/range_cache.yml
```

2. Use the prefix api path to make a request to query-frontend:

```
$ curl "http://localhost:9093/thanos/api/v1/query_range?query=time()&dedup=true&partial_response=true&start=1600269819.107&end=1600273419.107&step=14&max_source_resolution=0s&_=1600273386626"
```

3. Observe incrementing `http_requests_*` and `http_responses_*` metrics while `querier_*` and `cortex_*` metrics remain at 0 on localhost:9093/metrics

**Anything else we need to know**:

Looking at code https://github.com/thanos-io/thanos/blob/1078fe701ac2c58c06342de91dad23ca7963728a/pkg/queryfrontend/roundtrip.go#L109 it appears that the query-frontend has a strict path matcher which fails when downstream query has a path prefix:

```
			switch r.URL.Path {
			case "/api/v1/query":
...
			case "/api/v1/query_range":
...
			default:
			}
```

**Comment by yeya24** [comment]
Thanks for opening this issue. 

Currently query-frontend doesn't handle web route prefixes in the downstream querier.

Considering that we might want to configure multiple queriers with different web route, can we just check?

``` go
strings.HasSuffix(r.URL.Path, "/api/v1/query")
```

**Comment by gburek-fastly** [comment]
I think checking the suffix would be ideal and allow flexibility in routing to multiple downstreams.

**Comment by yeya24** [comment]
Fixed in #3195
